controllers/hvpa_controller_test.go

```diff
@@ -268,293 +285,342 @@ var _ = Describe("#TestReconcile", func() {
-					Expect(len(*blockedScaling)).To(Equal(len(data.expect.blockedReasons)))
 					for i, blockedScaling := range *blockedScaling {
 						Expect(blockedScaling.Reason).To(Equal(data.expect.blockedReasons[i]))
 					}
 				}
 
+				if data.expect.desiredReplicas == *target.Spec.Replicas && data.expect.resourceChange == false {
+					Expect(scaledStatus).To(BeNil())
+				} else {
+					Expect(scaledStatus.HpaStatus.DesiredReplicas).To(Equal(data.expect.desiredReplicas))
+				}
 				if data.expect.resourceChange {
-					Expect(podSpec).NotTo(BeNil())
-					Expect(podSpec.Containers[0].Resources).To(Equal(data.expect.resources))
+					Expect(newPodSpec).NotTo(BeNil())
+					Expect(newPodSpec.Containers[0].Resources).To(Equal(data.expect.resources))
 				} else {
-					Expect(podSpec).To(BeNil())
+					Expect(newPodSpec).To(BeNil())
 				}
-
-				hpaStatus, err = getWeightedReplicas(hpaStatus, hvpa, *target.Spec.Replicas, 100-vpaWeight, blockedScaling)
-				Expect(err).ToNot(HaveOccurred())
-				Expect(hpaStatus.DesiredReplicas).To(Equal(data.expect.desiredReplicas))
 			},
 
-			Entry("UpdateMode Auto, Should Scale only memory", &data{
+			Entry("UpdateMode Auto, scale up, paradoxical scaling, replicas increases, resources per replica decreases", &data{
 				setup: setup{
-					hvpa: newHvpa("hvpa-2", target.GetName(), "label-2", minChange),
-					hpaStatus: newHpaStatus(
-						2, 3, []autoscaling.HorizontalPodAutoscalerCondition{
-							{
-								Type:   autoscaling.ScalingLimited,
-								Status: v1.ConditionTrue,
-							},
-						}),
-					vpaStatus: newVpaStatus("deployment", "3G", "500m"),
-					target:    target,
-					vpaWeight: autoscalingv1alpha1.VpaWeight(40),
+					hvpa:      newHvpa("hvpa-2", target.GetName(), "label-2", minChange),
+					hpaStatus: nil,
+					vpaStatus: newVpaStatus("deployment", "2.2G", "150m"),
+					target:    newTarget("deployment", unscaledSmall, 1),
+				},
+				action: action{
+					limitScaling: limitScale,
+				},
+				expect: expect{
+					desiredReplicas: 2,
+					resourceChange:  false,
+					resources:       unscaledSmall,
+					blockedReasons:  []hvpav1alpha1.BlockingReason{},
+				},
+			}),
+			Entry("UpdateMode Auto, scaled down, no scaling because of paradoxical scaling recommendations", &data{
+				setup: setup{
+					hvpa:      newHvpa("hvpa-2", target.GetName(), "label-2", minChange),
+					hpaStatus: nil,
+					vpaStatus: newVpaStatus("deployment", "1.8G", "150m"),
+					target:    newTarget("deployment", unscaledLarge, 3),
 				},
 				action: action{
 					limitScaling: limitScale,
 				},
 				expect: expect{
-					scalingOff:      false,
 					desiredReplicas: 3,
+					resourceChange:  false,
+					blockedReasons: []hvpav1alpha1.BlockingReason{
+						hvpav1alpha1.BlockingReasonParadoxicalScaling,
+					},
+				},
+			}),
+			Entry("UpdateMode Auto, overall scale up, paradoxical scaling, but replicas decrease - should not be considered paradoxical", &data{
```

Comment:
If I understood right, this case is where the current replicas and requests are such that they are not within the right effective scaling interval for the `currentReplicas` and the recommendation results in a scale up but the effective scaling interval has the `desiredReplicas < currentReplicas`. Is that correct?

Reply:
Related to this scenario, could we also cover change in scaling interval spec such that the current replicas and requests that were valid for the effective scaling interval earlier but are now no longer valid after the spec change (for no change in recommendations, scale up and down)?

Reply:
>Is that correct?

Yes

> Related to this scenario, could we also cover change in scaling interval spec such that the current replicas and requests that were valid for the effective scaling interval earlier but are now no longer valid after the spec change (for no change in recommendations, scale up and down)?

will add in next commit